packages/dds/merge-tree/src/test/partialLength.spec.ts

```diff
@@ -0,0 +1,329 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { strict as assert } from "assert";
+import { UnassignedSequenceNumber } from "../constants";
+import { MergeTree } from "../mergeTree";
+import { MergeTreeDeltaType } from "../ops";
+import { PartialSequenceLengths } from "../partialLengths";
+import { TextSegment } from "../textSegment";
+import { insertText } from "./testUtils";
+
+describe("partial lengths", () => {
+    let mergeTree: MergeTree;
+    const localClientId = 17;
+    const remoteClientId = 18;
+
+    function getPartialLengths(
+        clientId: number,
+        seq: number,
+        mergeBlock = mergeTree.root,
+    ) {
+        const partialLen = mergeBlock.partialLengths?.getPartialLength(
+            seq,
+            clientId,
+        );
+
+        let actualLen = 0;
+
+        mergeTree.walkAllSegments(mergeBlock, (segment) => {
+            if (
+                segment.isLeaf()
+                && !(segment.removedSeq !== undefined && segment.removedSeq >= seq)
+                && segment.localRemovedSeq === undefined
+            ) {
+                actualLen += segment.cachedLength;
+            }
+            return true;
+        });
+
+        return {
+            partialLen,
+            actualLen,
+        };
+    }
+
+    function validatePartialLengths(
```

Comment:
I think these tests would be more comprehensive and easier to read if this function validated partial length results for all (clientId, seq) values in the collab window. You could make it take in an optional list of "known" values, but the rest can just be compared against the walkAllSegments result.